### PR TITLE
[SPARK-23941][MESOS] Mesos task failed on specific spark app name

### PR DESCRIPTION
**Pull request by samvantran**
Port from [SPARK#21014](https://github.com/apache/spark/pull/21014)

** edit: not a direct port from upstream Spark. Changes were needed because we saw PySpark jobs fail to launch when 1) run with docker and 2) including `--py-files`

## What changes were proposed in this pull request?
Shell escaped the name passed to spark-submit and change how conf attributes are shell escaped.

## How was this patch tested?
This test has been tested manually with Hive-on-spark with mesos or with the use case described in the issue with the sparkPi application with a custom name which contains illegal shell characters.

With this PR, hive-on-spark on mesos works like a charm with hive 3.0.0-SNAPSHOT.

I state that this contribution is my original work and that I license the work to the project under the project’s open source license

Author: Bounkong Khamphousone <bounkong.khamphousone@ebiznext.com>

Closes #21014 from tiboun/fix/SPARK-23941.

### CODE REVIEW

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 1]
```diff
@@ -542,13 +542,23 @@ private[spark] class MesosClusterScheduler(
...
       .toMap
     (defaultConf ++ driverConf).foreach { case (key, value) =>
-      options ++= Seq("--conf", s""""$key=${shellEscape(value)}"""".stripMargin) }
-
-    options
+      options ++= Seq("--conf", s"${key}=${value}") }
+    println("L554: options -> printing all options after key=value set")
+    println("================================================================")
+
+    val opts = options.map(shellEscape)
```
Comment (tnachen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Can we add a test for this? Also please remove the printlns

Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Was just a test, planning to delete the last commit

Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Doesn't look like Apache/Spark caught the issue for `--py-files`. Change+test added in [9f110fd](https://github.com/mesosphere/spark/pull/33/commits/9f110fd465b43cf255dc2b6306391ecf0ea19eb5)

**Comment by susanxhuynh** [comment]
@samvantran You might want to consider adding better unit test coverage of this code to prevent regressions in the future. (maybe in the upstream PR?) For example, (1) the case where the app name requires shell escaping and (2) the driverExtraJavaOptions with spaces. Any future changes to this code should handle these cases, in addition to the `--pyfiles` one that you did add a unit test for.

**Comment by samvantran** [comment]
Good point. I added a few more tests to this PR in [ef3e328](https://github.com/mesosphere/spark/pull/33/commits/ef3e32852ae2cf40f6b72f0be0106d93c12b61c6) but once we release, I'll create a Spark JIRA and PR to see if we can merge these upstream.

**Pull request merged**